### Tempus Interview: Case Study
#### Submitted by Nyssa Bulkes (20190614)
##### Instructions: 
The assignment is to investigate and analyze the attached data (DScasestudy.txt.gz), producing
a model that can be used to predict the “response”.
The first column of the provided data is the binary variable “response”. The 16,562 other
columns are binary columns that can be used to predict the “response”.


In [1]:
# import necessary modules
import pandas as pd
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split

In [2]:
# read in data: data
data = pd.read_csv('C:/Users/Megabean/Tempus/DScasestudy (1).txt.gz', compression='gzip', header=0, sep='\t', quotechar='"')
data.head()

response  V1  V2  V3  V4  V5  V6  V7  V8  V9   ...    V16553  V16554  \
0         1   0   0   0   0   0   1   0   0   0   ...         1       0   
1         1   0   0   0   0   0   1   0   0   0   ...         0       0   
2         1   0   0   0   0   0   1   0   0   0   ...         0       0   
3         1   0   0   0   0   0   1   0   0   0   ...         0       0   
4         1   0   0   0   0   0   1   0   0   0   ...         0       0   

   V16555  V16556  V16557  V16558  V16559  V16560  V16561  V16562  
0       0       0       0       0       0       0       0       0  
1       0       0       0       0       0       0       0       1  
2       0       0       0       0       0       0       0       1  
3       0       0       0       0       0       0       0       0  
4       0       0       0       0       0       0       0       0  

[5 rows x 16563 columns]

In [3]:
# take basic look at the data
data.describe()

response          V1          V2          V3     V4          V5  \
count  530.000000  530.000000  530.000000  530.000000  530.0  530.000000   
mean     0.232075    0.009434    0.009434    0.001887    0.0    0.020755   
std      0.422556    0.096761    0.096761    0.043437    0.0    0.142697   
min      0.000000    0.000000    0.000000    0.000000    0.0    0.000000   
25%      0.000000    0.000000    0.000000    0.000000    0.0    0.000000   
50%      0.000000    0.000000    0.000000    0.000000    0.0    0.000000   
75%      0.000000    0.000000    0.000000    0.000000    0.0    0.000000   
max      1.000000    1.000000    1.000000    1.000000    0.0    1.000000   

          V6          V7          V8     V9     ...          V16553  V16554  \
count  530.0  530.000000  530.000000  530.0     ...      530.000000   530.0   
mean     1.0    0.001887    0.030189    0.0     ...        0.158491     0.0   
std      0.0    0.043437    0.171268    0.0     ...        0.365545     0.0   
min      1.0    0.000000    0.000000    0.0     ...        0.000000     0.0   
25%      1.0    0.000000    0.000000    0.0     ...        0.000000     0.0   
50%      1.0    0.000000    0.000000    0.0     ...        0.000000     0.0   
75%      1.0    0.000000    0.000000    0.0     ...        0.000000     0.0   
max      1.0    1.000000    1.000000    0.0     ...        1.000000     0.0   

       V16555      V16556      V16557      V16558  V16559      V16560  V16561  \
count   530.0  530.000000  530.000000  530.000000   530.0  530.000000   530.0   
mean      0.0    0.001887    0.020755    0.013208     0.0    0.003774     0.0   
std       0.0    0.043437    0.142697    0.114270     0.0    0.061371     0.0   
min       0.0    0.000000    0.000000    0.000000     0.0    0.000000     0.0   
25%       0.0    0.000000    0.000000    0.000000     0.0    0.000000     0.0   
50%       0.0    0.000000    0.000000    0.000000     0.0    0.000000     0.0   
75%       0.0    0.000000    0.000000    0.000000     0.0    0.000000     0.0   
max       0.0    1.000000    1.000000    1.000000     0.0    1.000000     0.0   

           V16562  
count  530.000000  
mean     0.447170  
std      0.497671  
min      0.000000  
25%      0.000000  
50%      0.000000  
75%      1.000000  
max      1.000000  

[8 rows x 16563 columns]

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 530 entries, 0 to 529
Columns: 16563 entries, response to V16562
dtypes: int64(16563)
memory usage: 67.0 MB


Calling the describe and info methods above shows: 1) the data contains 530 unique observations and over 16000 columns; and 2) that none of the observations are null values. We can skip missing-data handling.

The instructions of the assignment suggest the first column is the response variable, and that all of the observations are binary. If the goal is to predict the outcome measure, we can implement a simple approach to predict the dependent measure, <b>namely a logistic regression</b> (due to the binary nature of the data).  

In [5]:
# Identify feature values
X = data.iloc[:,1:]

# Identify response variable, "response"
y = data.iloc[:,0]

In [6]:
# Split the data into training, testing sets
Xlr, Xtestlr, ylr, ytestlr = train_test_split(X.values, y.values,random_state=5)

# Instantiate the model
model = LogisticRegression()

# Train the model
model.fit(Xlr, ylr)

# Print the accuracy using the test data
print(accuracy_score(model.predict(Xtestlr), ytestlr))

0.8947368421052632


Above, the data is split into training and testing datasets and a model instantiated. Once the model is trained, on the testing data, the model performs with approximately 89.47% accuracy. Namely, this suggests the model can predict the response variable on the unseen testing data about 89% of the time.

Follow-up steps to this basic instantiation could be to tune the hyperparameters to ensure we're getting the best model fit possible. Further, we could entertain other possible algorithms to model the data, as there isn't likely just one correct answer. For example, knowing that there are two possibilities in the response variable, another classifier might be appropriate (i.e. KNN), but without more knowledge of what the desired outcome of this model is, a logistic regression seems to be the simplest solution.